src/frontend/js/hooks/useResources/useResourcesRoot.ts

```diff
@@ -0,0 +1,182 @@
+  NonNullable<TApiMethod>,
+  [options?: MutateOptions<Awaited<ReturnType<NonNullable<TApiMethod>>>, HttpError>]
+>;
+
+export const useResourcesRoot = <
+  TData extends Resource,
+  TResourceQuery extends ResourcesQuery = ResourcesQuery,
+  TApiResource extends ApiResourceInterface<TData> = ApiResourceInterface<TData>,
+>({
+  queryKey,
+  apiInterface,
+  frozenQueryKey,
+  filters,
+  session,
+  queryOptions,
+  localized,
+  messages: resourceMessages,
+}: UseResourcesProps<TData, TResourceQuery, TApiResource>) => {
+  const queryClient = useQueryClient();
+
+  const [error, setError] = useState<Maybe<string>>();
+  const intl = useIntl();
+  const COMMON_QUERY_KEY = frozenQueryKey ? [...queryKey] : [...queryKey, JSON.stringify(filters)];
+  const LOCALIZED_QUERY_KEY = useLocalizedQueryKey(COMMON_QUERY_KEY);
+  const QUERY_KEY = localized ? LOCALIZED_QUERY_KEY : COMMON_QUERY_KEY;
+  let ACTUAL_QUERY_KEY = QUERY_KEY;
+  const actualMessages = { ...messages, ...resourceMessages };
+  const api = apiInterface();
+
+  const queryFn: () => Promise<any> = () => api.get(filters);
+  queryOptions = {
+    onError: () => setError(intl.formatMessage(actualMessages.errorSelect)),
+    ...queryOptions,
+  };
+
+  if (session !== usePrevious(session)) {
+    throw new Error('session must never change value.');
+  }
+
+  let readHandler: UseQueryResult<any, HttpError>;
+  if (session) {
+    [readHandler, ACTUAL_QUERY_KEY] = useSessionQuery(QUERY_KEY, queryFn, queryOptions as any);
+  } else {
+    readHandler = useQuery(QUERY_KEY, queryFn, queryOptions);
+  }
+
+  const invalidate = async () => {
+    // Invalidate all queries related to the resource
+    await queryClient.invalidateQueries({
+      predicate: (query) => query.queryKey.includes(queryKey[0]),
+    });
+  };
+
+  const prefetch = async () => {
+    await queryClient.prefetchQuery(ACTUAL_QUERY_KEY, queryFn, {
+      staleTime: REACT_QUERY_SETTINGS.staleTimes.sessionItems,
+    });
+  };
+
+  const onSuccess = async () => {
+    setError(undefined);
+    await invalidate();
+  };
+
+  const writeHandlers = {
+    create: api.create
+      ? useSessionMutation(api.create, {
+          onSuccess,
+          onError: () => setError(intl.formatMessage(actualMessages.errorCreate)),
+        })
+      : undefined,
+    update: api.update
+      ? useSessionMutation<TData, any>(api.update, {
+          onSuccess,
+          onError: () => setError(intl.formatMessage(actualMessages.errorUpdate)),
+        })
+      : undefined,
+    delete: api.delete
+      ? useSessionMutation<void, string>(api.delete, {
+          onSuccess,
+          onError: () => setError(intl.formatMessage(actualMessages.errorDelete)),
+        })
+      : undefined,
+  };
+
+  // We want to keep the same reference to the empty array to avoid potential
+  // infinite useEffect calls that use `items` as a dependency.
+  const emptyArray = useMemo(() => [], []);
```

Comment:
I just wonder what is the most efficient between `useRef` and `useMemo` for this case 🤔

Reply:
Check-out the fixup I made something 100% more efficient :p

Reply:
But now `emptyArray` is a global variable so shared across hook instances.

Reply:
Indeed better and simpler ^^